README.md
This is the NSL (National Species List) Services Application developed at the Australian National Botanic Gardens in
association with CHAH, the ALA and the IBIS team.

The NSL services provide a relatively generic set of services for searching Names, Refererences and Authors as well as
accessing classification trees of taxons.

The NSL is (and probably always will be) a work in progress, we'll be adding a roadmap and more doco here as we go.

This code is Licensed under the Apache 2 License and we welcome contributions from the taxonomic community.

Please use Github issues for any bug reports.

Shard ID ranges
====

For Australian NSL infrastructure we are splitting the ID ranges used across different known shards. This is *not* essential
but it does mean we have a means to discriminate data sources on IDs if somehow we manage to loose track.

The current allocated ranges are:

* Vascular (APNI) 1000 - 10,000,000 + 50,000,001 - 60,000,000
* Moss 10,000,001 - 20,000,000
* Algae 20,000,001 - 30,000,000
* Lichen 30,000,001 - 40,000,000
* DAWR 40,000,001 - 50,000,000 (Aust. Dept. of Agriculture and Water Resources)
* Fungi 60,000,001 - 70,000,000
* AFD 70,000,001 - 80,000,000

There is a confluence page for the NSL infrastructure as well at https://www.anbg.gov.au/ibis25/pages/viewpage.action?spaceKey=NSL&title=NSL+Project+2.0


docs/changelog.adoc
= Changelog
:imagesdir: resources/images/
:toc: left
:toclevels: 4
:toc-class: toc2
:icons: font
:iconfont-cdn: //cdnjs.cloudflare.com/ajax/libs/font-awesome/4.3.0/css/font-awesome.min.css
:stylesdir: resources/style/
:stylesheet: asciidoctor.css
:description: Services Configuration
:keywords: documentation, NSL, change log
:links:
:numbered:

== v1.0235
* NSL-3831: Fixed issues in last version of synonymy changes
* Changes to audit reporting
* Added more admin functions
* Changes 1753 job to update only outdated instances
* Fix concurrency issues with cleanUp job

== v 1.0234
* Add index action to search controller
* Added afd shard assets and styles
* NSL-3831: First draft tree synonymy auto-update functionality for author and reference
* NSL-3845: Fix ability to delete comments and distribution from services
* Fix concurrency issues with grails jobs
* Changes to Search endpoint to avoid logs from filling up (WIP)

== v 1.0233
* Fixed font awesome icons for protologue and images
* Fixed classification in APC view to be a popup
* Fix error logging from going to STDOUT instead of log file
* NSL-3811: Services will now re-synchronize with mapper token
* NSL-3827: Added a new button and collapsible search interface to name check results screen
* NSL-3844: Fixed arrows in datepicker. Setup logging based on environment
* Fix the classification to be a popup instead of displaying online; Need to fix new line characters
* Fixed bugs in Tree saves
* Added additional admin functions to run from services
* Fix names and links on exports tabs
* Add validations to run db functions from admin screen
* Change read timeouts to work on cloud
* NSL-3989: Fix Tree distribution entries not saving
* NSL-4003: Optimize tree distribution

== v 1.0232
* NSL-3661 Fixed the broken tree and taxon URLs from editor to services
* NSL-3659: Renamed the view name to match controller action
* NSL-3702: Elements missing on tree action; Fix padding and removed box layout
* NSL-3705: Fixed the icons in the timeline
* NSL-3589: Swapped the unicode character ⇅ with fa icon
* NSL-3705: Fixed the icons in the timeline
* NSL-3589: Swapped the unicode character ⇅ with fa icon
* NSL-3729: Fixed the role to match for deleting comments and distribution from tree elements
* NSL-3738: Fixed the json response for taxon-search to have accepted name and synonyms
* NSL-3738: Changed response test in test cases for author and name
* NSL-3704: Change the sorting in tree index to be by published_date
* Get broadcasts file to be recognised in g3
* Fixed the app version to display in Dashboard
* Removed the /boa/foa/taxa/ calls from being sent to a non-existent end-point
* Change logback log pattern
* Fix ajax event preventDefault();
* Increase the timeout duration for services to wait
* NSL-3792: Dashes appearing for no-author works during reference citation reconstruction

== v 1.0231

* Docker security changes.
* NSL-3544: Modify RestResourceController.node action to use the taxon code.
* NSL-3549: Add admin script to fix broken name paths
* NSL-3661: Fixed the broken tree and taxon URLs from editor to services
* NSL-3659: Renamed the view name to match controller action
* NSL-3702: Elements missing on tree action; Fix padding and removed box layout
* NSL-3705: Fixed the icons in the timeline
* NSL-3589: Swapped the unicode character ⇅ with fa icon
* NSL-3729: Fixed the role to match for deleting comments and distribution from tree elements
* NSL-3738: Fixed the json response for taxon-search to have accepted name and synonyms
* NSL-3738: Changed response test in test cases for author and name
* Get broadcast messages working in g3 services
* Fixed the app version to display in the *Dashboard* tab
* NSL-3704: Change the sorting in tree index to be by published_date

== v 1.0229

* NSL-3453: show the latest _published_ accepted taxonomy, not draft accepted taxonomy, for a taxon link.

== v 1.0226

* Fix API and Auth interceptor tests
* NSL-3328 format isoPublicationDate for citations (cross port)
* Update to use the new mapper-mn with JWT and changes to API 
* *Removed the link cache*
* Documentation updates. 
 ** Added change log 
 ** Added config.adoc 
 ** Added docker.adoc 
* NSL-3356 changes to get diff output working again.

== v 1.0225

* NSL-3356 improve diff output to make it easier to see changes Increase the number of synonym changes shown on screen
to 100 (which may be reverted). 
* Fix search on apni-format and apc-format pages Add a docker file to build docker image Make the config file
configurable via an environment variable 
* NSL-3356 add an ordering report for synonymy on an instance that you can get from the check synonymy report. Add back
the moved indicator on the synonymy diff output. 
* Change nsl-domain plugin version to 3.1.19-SNAPSHOT to determine the grails version of the plugin to use. 
* set the version to the G2 version of the services. 
* update the nsl-domain-plugin to 1.19.SNAPSHOT due to change in the domain plugin. 

== v 1.0224 - initial grails 3 version

* Clean up audit UI and move to grails-shiro 3.1 release 
* Add readme 
* Fix classifications page layout. 
* Make the runAs user check the roles of the user being runAs Make gradle give us enough memory. 
* Get basic API realm going again. 
* Improve the content display area and keep the header constant. 
* Fix NSL-3348 and move to domain plugin 1.19-SNAPSHOT 
* Make dashboard collapse on summary data work again 
* Clean up name check page. make search pages and heading not jump when reloading 
* Clean up make consistent api calls 
* Fix some broken links. Fix missing injected dependency in NameControllerSpec 
* Port changes from grails 2 master. 
* Clean up 
* Further improve the admin log output. replace all services usage of runAsync 
* minor doc fix 
* Fix admin page log pain by finding the correct logger. Add log paths to the configuration tab. Refactor the name service

etc/nsl/editor-config.rb
#hosts
external_services_host = "http://#{ENV['EXT_HOST']}:8080/nsl/services"
internal_services_host = "http://#{ENV['EXT_HOST']}:8080/nsl/services"
internal_mapper_host = "http://#{ENV['EXT_HOST']}:7070"
external_mapper_host = "http://#{ENV['EXT_MAPPER_HOST']}"

#environment
Rails.configuration.action_controller.relative_url_root = "/nsl/editor"
Rails.configuration.environment = 'production'
Rails.configuration.session_key_tag = 'prod'
Rails.configuration.draft_instances = 'true'

#Services
Rails.configuration.services_clientside_root_url = "#{external_services_host}/"
Rails.configuration.services = "#{internal_services_host}/"
#TODO deprecate these
Rails.configuration.name_services = "#{internal_services_host}/rest/name/apni/"
Rails.configuration.reference_services = "#{internal_services_host}/rest/reference/apni/"

#used to create external facing links to the services
Rails.configuration.nsl_links = "#{external_services_host}/"

#API key for the services
Rails.configuration.api_key = 'dev-apni-editor'

#mapper                                                        
Rails.configuration.x.mapper_api.version = 2
Rails.configuration.x.mapper_api.url = "#{internal_mapper_host}/api/"
Rails.configuration.x.mapper_api.username = 'TEST-services'
Rails.configuration.x.mapper_api.password = 'buy-me-a-pony'
Rails.configuration.x.mapper_external.url = "#{external_mapper_host}/"

#LDAP
Rails.configuration.ldap_admin_username = "uid=admin,ou=system"
Rails.configuration.ldap_admin_password = "secret"
Rails.configuration.ldap_base = "ou=users,o=shards"
Rails.configuration.ldap_host = "#{ENV['EXT_HOST']}"
Rails.configuration.ldap_port = 10389
Rails.configuration.ldap_users = "ou=users,o=shards"
Rails.configuration.ldap_groups = "ou=groups,ou=#{ENV['SHARD']},o=shards"


grails-app/assets/javascripts/apniFormat.js
/**
 * Created by pmcneil on 17/09/14.
 * Javascript mainly to do with the apniFormat displays
 * todo maybe re-name to format.js and split out the apninFormat specific bits
 */
$(function () {
  if (typeof internetExplorer !== 'undefined') {
    $.ajaxSetup({cache: false});
  }
  var lazyLoad = function () {
    var delay = 0;
    $('div.results > div > div.unfetched.name').each(function () {
        var el = $(this);
        var getName = function () {
          var id = el.attr('id');
          var format = el.data('format');
          var product = el.data('product');
          if (format != 'noneFormat') {
            var url = format + "/name/" + id + "?product=" + product;
            window.console && console.log("found a name, getting " + url);
            $.get(url, function (data, status, request) {
              window.console && console.log('status ' + status + ', request ' + request);
              if (data !== undefined) {
                var parentDiv = el.parent('div');
                el.replaceWith(data);
                $('.toggleNext').unbind('click').click(function () {
                  toggleNext(this);
                });
                parentDiv.find("protologue-pdf").each(function () {
                  checkProtologue(this);
                });
                parentDiv.find('branch').click(function (event) {

                  if ($(event.target).is("a")) {
                    // default behaviour on a hyperlink click
                  }
                  else {
                    $(this).children('ul').toggleClass("pop-classify");
                    event.preventDefault();
                  }
                });
                parentDiv.find('date').each(function () {
                  var d = $(this).html();
                  $(this).html(jQuery.format.prettyDate(d));
                  $(this).html(jQuery.format.prettyDate(d));
                });
              }
            });
          }
        };
        window.setTimeout(getName, delay);
        delay = delay + 100;
      }
    )
    ;
  };
  window.setTimeout(lazyLoad, 100);

  $('.toggleNext').unbind('click').click(function () {
    toggleNext(this);
  });

  var toggleNext = function (el) {
    $(el).find('i').toggle();
    $(el).next('div').toggle(200);
  };

  $('.toggleNextRow').unbind('click').click(function () {
    toggleNextRow(this);
  });

  var toggleNextRow = function (el) {
    $(el).find('i').toggle();
    $(el).next('tr').toggle(200);
  };

  $(".closeForm").click(function () {
    $('.openForm').show();
    $('.closeForm').hide();
    $('.panel-body > form.closable').toggle();
  });

  $(".openForm").click(function () {
    $('.openForm').hide();
    $('.closeForm').show();
    $('.panel-body > form.closable').toggle();
  });

  $(".hideSearch").each(function () {
    $('.closeForm').addClass('highlight').focus();

    window.setTimeout(function () {
      $('.openForm').show();
      $('.closeForm').hide();
      $('.closeForm').removeClass('highlight');
      $('.panel-body > form.closable').toggle(200);
    }, 1000);
  });

  $("#expandAll").click(function () {
    //window.setTimeout(lazyLoad, 200);
    $('family').show();
    $('.instances').show();
    $('.fa-caret-down.toggleNext').hide();
    $('.fa-caret-up.toggleNext').show();
  });

  $("#collapseAll").click(function () {
    $('family').hide();
    $('.instances').hide();
    $('.fa-caret-down.toggleNext').show();
    $('.fa-caret-up.toggleNext').hide();
  });

  $(".loadFormat").click(function () {
    var url = $(this).attr('href');
    var parent = $(this).parent('div').parent('div').parent('div');
    $.get(url, function (data, status, request) {
      window.console && console.log('status ' + status + ', request ' + request);
      if (data != undefined) {
        parent.html(data);
      }
      $('.toggleNext').unbind('click').click(function () {
        toggleNext(this);
      });
    });
    return false;
  });

  $(".suggest").each(function () {
    var action = $(this).data('subject');
    var context = $(this).data('context');
    var quoted = $(this).data('quoted');
    var actionurl = function (request, response) {
      var url = baseContextPath + '/suggest/' + action + '?term=' + encodeURIComponent(request.term);
      if (context != undefined) {
        var contextElement = $("#" + context);
        var contextValue = contextElement.val();
        url += '&context=' + encodeURIComponent(contextValue);
      }
      $.get(url, function (data, status, request) {
        response(data);
      });
    };
    $(this).autocomplete({
      minLength: 1,
      source: actionurl,
      select: function (event, ui) {
        var qry = ui.item.value;
        if (quoted) {
          qry = '"' + ui.item.value + '"';
        }
        if (qry != '...') {
          $(this).val(qry);
        }
        event.preventDefault();
      }
    });
  });

  $(".checkbig").submit(function () {
    var data = $(this).serialize();
    if (data.length > 4096) {
      $(this).attr('method', 'POST');
    }
  });

  $("foa").each(function () {
    var url = $(this).data('id');
    window.console && console.log("found foa, getting " + url);
    $.get(url, function (data, status, request) {
      window.console && console.log('status ' + status + ', request ' + request);
      if (data != undefined) {
        var html = $.parseHTML(data);
        var content = $(html).find('div.foa-content');
        var footer = $(html).find('div.foa-footer');
        $("foa").append(content);
        $("div.foa-content").append(footer);
        $("#foaToggle").show();
      }
    });
  });

  var checkProtologue = function (el) {
    var thisEl = el;
    var url = $(el).data('id');
    var link = '<a title="Protologue PDF image" href="' + url + '"><i class="far fa-file-pdf"></i></a>';
    window.console && console.log("found protologue pdf, trying " + url);
    $.ajax({
      type: "HEAD",
      async: true,
      url: url,
      success: function () {
        window.console && console.log("protologue pdf exists");
        $(thisEl).append(link);
      }
    });
  };

  $("protologue-pdf").each(function () {
    checkProtologue(this);
  });

  $("#fontToggle").click(function (event) {
    var results = $('.results');
    var font = results.css("font-family");
    if (font.startsWith("\"Book")) {
      results.css("font-family", "\"Lucida Grande\", \"Helvetica Nueue\", Arial, sans-serif");
      $.cookie('resultFont', 'sans', {expires: 365});
    } else {
      results.css("font-family", "\"Bookman\", Georgia, \"Times New Roman\", serif");
      $.cookie('resultFont', 'serif', {expires: 365});
    }
    event.preventDefault();
  });

  $("#inRank")
    .each(function () {
      var el = $(this);
      if (el.val()) {
        $('#rankName').prop('disabled', false);
      } else {
        $('#rankName').val('').prop('disabled', true);
      }
    })
    .change(function (even) {
      var el = $(this);
      if (el.val()) {
        $('#rankName').prop('disabled', false).focus();
      } else {
        $('#rankName').val('').prop('disabled', true);
      }
    });

//check user preference for sans serif fonts on results
  var prefFont = $.cookie('resultFont');
  if (prefFont) {
    if (prefFont == 'sans') {
      $('.results').css("font-family", "\"Lucida Grande\", \"Helvetica Nueue\", Arial, sans-serif");
    }
  }

  $('branch').click(function (event) {

    if ($(event.target).is("a")) {
      // default behaviour on a hyperlink click
    }
    else {
      $(this).children('ul').toggle();
      event.preventDefault();
    }
  });
})
;

function clearComment(el) {
  $(el).parent().parent('form').find('[name="comment"]').val('');
}

function clearDistribution(el) {
  $(el).parent().parent('form').find('[name="distribution"]').val('');
}

function clearValue(el) {
  $(el).parent().parent('form').find('[name="value"]').val('');
}

function confirmDelete(el) {
  $(el).prop('disabled', true).next('div').show().focus();
  $(el).parent('form').find('[name="reason"]').val('Deleted, incorrect.');
}

function cancelDelete(el) {
  $(el).parent().hide().prev('button').prop('disabled', false);
  $(el).parent().parent('form').find('[name="reason"]').val('Errata, typographic error');
}



src/main/groovy/au/org/biodiversity/nsl/JsonWebToken.java
package au.org.biodiversity.nsl;

import io.jsonwebtoken.*;
import org.apache.shiro.authc.AuthenticationToken;

import java.security.Key;

public class JsonWebToken implements AuthenticationToken {
    private final String credentials;
    private final String principal;
    private final Jws<Claims> claims;

    public JsonWebToken(String jwt, Key key) throws ExpiredJwtException, UnsupportedJwtException, MalformedJwtException, SignatureException, IllegalArgumentException {
            this.credentials = jwt;
            this.claims = Jwts.parser()
                    .setSigningKey(key)
                    .parseClaimsJws(credentials);
            this.principal = claims.getBody().getSubject();
    }

    @SuppressWarnings("unused")
    public Jws<Claims> getClaims() {
        return claims;
    }

    @Override
    public Object getPrincipal() {
        return principal;
    }

    @Override
    public Object getCredentials() {
        return credentials;
    }
}


grails-app/assets/javascripts/application.js
// This is a manifest file that'll be compiled into application.js.
//
// Any JavaScript file within this directory can be referenced here using a relative path.
//
// You're free to add application-wide JavaScript to this file, but it's generally better 
// to create separate JavaScript files as needed.
//
//= require jquery.cookie
//= require jquery-dateformat.min
//= require apniFormat
//= require_self

var pendingReload = false;

function log(message) {
  window.console && console.log(message);
}

if (typeof jQuery !== 'undefined') {
  log("Yes we have JQuery");
  (function ($) {
    $('#spinner').ajaxStart(function () {
      $(this).fadeIn();
    }).ajaxStop(function () {
      $(this).fadeOut();
    });

  })(jQuery);
}

function editTreeData(event) {
  event.preventDefault();
  if (!pendingReload) {
    log("in edit tree data");
    var form = $("#treeDetailsForm");
    var data = {
      id: form.find('#treeId').val(),
      treeName: form.find('#treeName').val(),
      groupName: form.find('#groupName').val(),
      publication: form.find('#publication').val(),
      descriptionHtml: form.find('#descriptionHtml').val(),
      linkToHomePage: form.find('#linkToHomePage').val(),
      acceptedTree: form.find('#acceptedTree').prop("checked")
    };

    if (data.id && data.id !== "") {
      send(data, 'POST', baseContextPath + "/api/tree/editTree");
    } else {
      send(data, 'PUT', baseContextPath + "/api/tree/createTree");
    }
  }
}

function editDraftData(event) {
  event.preventDefault();
  if (!pendingReload) {
    var theForm = $(event.relatedTarget);
    log("in edit draft data");
    var form = $("#draftDetailsForm");
    var data = {
      treeId: form.find('#vtreeId').val(),
      versionId: form.find('#versionId').val(),
      fromVersionId: form.find('#fromVersionId').val(),
      draftName: form.find('#draftName').val(),
      defaultDraft: form.find('#defaultDraft').prop("checked")
    };

    if (data.versionId && data.versionId !== "") {
      send(data, "POST", baseContextPath + "/api/treeVersion/edit");
    } else {
      send(data, 'PUT', baseContextPath + "/api/tree/createVersion");
    }

  }
}

function publishDraftData(event) {
  event.preventDefault();
  if (!pendingReload) {
    log("in publish draft data");
    var form = $("#publishDraftForm");
    var data = {
      versionId: form.find('#draftVersionId').val(),
      logEntry: form.find('#logEntry').val()
    };

    send(data, "POST", baseContextPath + "/api/treeVersion/publish");
  }
}

function deleteDraftData(event) {
  event.preventDefault();
  if (!pendingReload) {
    log("in delete draft data");
    var form = $("#deleteDraftForm");
    var deleteUrl = form.find('#deleteUrl').val();

    disableForms();
    $.ajax({
      method: "POST",
      url: deleteUrl
    }).done(function () {
      location.reload(true);
    }).fail(function (jqxhr, statusText) {
      if (jqxhr) {
        if (jqxhr.status === 403) {
          log("status 403 forbidden");
          alert("Apparently you're not allowed to do that.");
        } else if (jqxhr.responseJSON) {
          log("Fail: " + statusText + ", " + jqxhr.responseJSON.error);
          alert('Failed to delete: ' + jqxhr.responseJSON.error);
        } else if (jqxhr.responseText) {
          log("Fail: " + statusText + ", " + jqxhr.responseText);
          alert("Failed to delete with status " + statusText + ". " + jqxhr.responseText);
        } else if (jqxhr.responseXML) {
          log("Fail: " + statusText + ", " + jqxhr.responseText);
          alert("Failed to delete with status " + statusText + ". XML:" + jqxhr.responseXML);
        }
      }
      enableForms();
    });
  }
}

function disableForms() {
  pendingReload = true;
  $('button').attr('disabled', 'disabled');
  $("button[type='submit'] i.fa").addClass('fa-spin');
}

function enableForms() {
  pendingReload = false;
  $('button').removeAttr('disabled');
  $("button[type='submit'] i.fa").removeClass('fa-spin');
}

function send(data, method, url) {
  disableForms();
  $.ajax({
    method: method,
    url: url,
    data: JSON.stringify(data),
    contentType: "application/json",
    dataType: "json"
  }).done(function () {
    location.reload(true);
  }).fail(function (jqxhr, statusText) {
    if (jqxhr.status === 403) {
      log("status 403 forbidden");
      alert("Apparently you're not allowed to do that.");
    } else if (jqxhr.responseJSON) {
      log("Fail: " + statusText + ", " + jqxhr.responseJSON.error);
      alert(jqxhr.responseJSON.error);
    } else if (jqxhr.responseText) {
      log("Fail: " + statusText + ", " + jqxhr.responseText);
      alert("That didn't work: " + statusText + ". " + jqxhr.responseText);
    }
    enableForms();
  });
}

function replaceDates() {
  $('date').each(function (element) {
    var d = $(this).html().toString();
    $(this).html(jQuery.format.prettyDate(d));
    $(this).attr('title', d);
  });
}

$(function () {

  function worker() {
    var logs = $('body').find('#logs').length;
    if (logs == 1) {
      $.ajax({
        url: 'logs',
        success: function (data, statusText, jqXHR) {
          //note redirects will be followed
          if (jqXHR.status == 200) {
            $('#logs').html(data);
            setTimeout(worker, 5000);
          } else {
            location.reload(true);
          }
        }
      });
    }
  }

  worker();


  $('#productDescription').on('closed.bs.alert', function () {
    var cookieName = 'close' + $(this).data('product') + 'Description';
    $.cookie(cookieName, 'true', {expires: 30});
  }).each(function () {
    var cookieName = 'close' + $(this).data('product') + 'Description';
    var closeProductDescription = $.cookie(cookieName);
    if (closeProductDescription) {
      $('#productDescription').alert('close');
    }
  });

  $('help').click(function (event) {
    $(this).children('div').toggle();
    event.preventDefault();
  });

  $('input.fromDate').datepicker({dateFormat: 'd/m/yy', defaultDate: '-1w', maxDate: 0});
  $('input.toDate').datepicker({dateFormat: 'd/m/yy', maxDate: 0});

  $('#editTree').on('show.bs.modal', function (event) {
    var button = $(event.relatedTarget);
    var name = button.data('name');
    var group = button.data('group');
    var ref = button.data('ref');
    var desc = button.data('desc');
    var link = button.data('link');
    var accepted = button.data('accepted');
    var treeId = button.data('tree');
    var title = button[0].title;

    var modal = $(this);
    modal.find('#treeId').val(treeId);
    modal.find('#treeName').val(name);
    modal.find('#groupName').val(group);
    modal.find('#publication').val(ref);
    modal.find('#descriptionHtml').val(desc);
    modal.find('#linkToHomePage').val(link);
    if (accepted) {
      modal.find('#acceptedTree').prop("checked", "checked");
      modal.find('#acceptedTree').attr("checked", "checked");
    } else {
      modal.find('#acceptedTree').removeAttr("checked");
    }
    modal.find('.modal-title').text(title);
  });


  $('#editDraft').on('show.bs.modal', function (event) {
    var button = $(event.relatedTarget);
    var name = button.data('name');
    var defaultDraft = button.data('default-draft');
    var treeId = button.data('tree-id');
    var versionId = button.data('version-id');
    var fromVersionId = button.data('from-version-id');
    var title = button[0].title;

    var modal = $(this);
    if (defaultDraft) {
      modal.find('#defaultDraft').prop("checked", "checked");
      modal.find('#defaultDraft').attr("checked", "checked");
    } else {
      modal.find('#defaultDraft').removeAttr("checked");
    }
    modal.find('#draftName').val(name);
    modal.find('#vtreeId').val(treeId);
    modal.find('#versionId').val(versionId);
    modal.find('#fromVersionId').val(fromVersionId);
    modal.find('.modal-title').text(title);
  });

  $('#publishDraft').on('show.bs.modal', function (event) {
    var button = $(event.relatedTarget);
    var versionId = button.data('version-id');
    var title = button[0].title;

    var modal = $(this);
    modal.find('#draftVersionId').val(versionId);
    modal.find('.modal-title').text(title);
  });

  $('#deleteDraft').on('show.bs.modal', function (event) {
    var button = $(event.relatedTarget);
    var url = button.data('url');
    var draftName = button.data('draft-name');
    var title = button[0].title;

    var modal = $(this);
    modal.find('#deleteUrl').val(url);
    modal.find('#deleteDraftName').text(draftName);
    modal.find('.modal-title').text(title);
  });


  $('#treeDetailsForm').on('submit', function (event) {
    editTreeData(event);
  });
  $('#draftDetailsForm').on('submit', function (event) {
    editDraftData(event);
  });
  $('#publishDraftForm').on('submit', function (event) {
    publishDraftData(event);
  });
  $('#deleteDraftForm').on('submit', function (event) {
    deleteDraftData(event);
  });

  replaceDates();

  // merge stuff
  $('div.tve-diff-from label').each(function () {
    var parentid = $(this).data('parentid');
    if (parentid !== undefined && parentid !== "") {
      $(this).addClass('disabled').find('input').attr('disabled', 'disabled');
    }
    $(this).find('input').on('click', function () {
      var parentid = $(this).data('diffid');
      console.log('Clicked from ' + parentid);
      var sel = 'div.tve-diff-from label[data-parentid="' + parentid + '"]';
      $(sel).removeClass('disabled').find('input').removeAttr('disabled').click();
    });
  });

  $('div.tve-diff-to label input').on('click', function () {
    var parentid = $(this).data('diffid');
    console.log('Clicked to ' + parentid);
    $('div.tve-diff-from label[data-parentid="' + parentid + '"]')
      .addClass('disabled')
      .find('input')
      .attr('disabled', 'disabled');
    $('div.tve-diff-to label[data-parentid="' + parentid + '"] input').click();
  });


});
